src/out_of_order_engine.cc

```diff
@@ -0,0 +1,430 @@
+			auto& successor = succ_it->second;
+			assert(successor.num_incomplete_predecessors > 0);
+			--successor.num_incomplete_predecessors;
+			try_mark_for_assignment(successor);
+		}
+	}
+}
+
+bool engine_impl::is_idle() const {
+#ifndef NDEBUG
+	if(incomplete_instructions.empty()) {
+		assert(assignment_queue.empty());
+		for(const auto& device_state : device_queue_target_states) {
+			for(const auto& lane : device_state.lanes) {
+				assert(lane.num_in_flight_assigned_instructions == 0);
+			}
+		}
+		for(const auto& lane : host_queue_target_state.lanes) {
+			assert(lane.num_in_flight_assigned_instructions == 0);
+		}
+	}
+#endif
+	return incomplete_instructions.empty();
+}
+
+incomplete_instruction_state* engine_impl::pop_assignable() {
+	while(!assignment_queue.empty()) {
+		const auto instr = assignment_queue.top();
+		assignment_queue.pop();
+
+		auto& node = incomplete_instructions.at(instr->get_id());
+		assert(node.num_unassigned_predecessors == 0);
+
+		if(const auto eagerly_assignable_when_pushed = std::get_if<conditional_eagerly_assignable_state>(&node.assignment)) {
+			assert(node.num_incomplete_predecessors > 0); // otherwise this would be an immediately_assignable_state
+			const auto& lane = get_lane_state(node.target, eagerly_assignable_when_pushed->device, eagerly_assignable_when_pushed->lane);
+			if(lane.last_incomplete_submission == nullptr
+			    || lane.last_incomplete_submission == eagerly_assignable_when_pushed->expected_last_submission_on_lane) {
+				// Our preferred lane is still in the required state to go through with eager assignment
+				return &node;
+			} else {
+				// A third instruction has been submitted to our preferred lane since, so we drop eager assignment and don't need to attempt it again, since
+				// none of our dependencies can now end up last in the queue anymore.
+				node.assignment.emplace<unassigned_state>().probe_for_eager_assignment = false;
+				continue;
+			}
+		} else {
+			assert(std::holds_alternative<unconditional_assignable_state>(node.assignment));
+			assert(node.num_incomplete_predecessors == 0);
+			return &node;
+		}
+	}
+
+	return nullptr;
+}
+
+std::optional<assignment> engine_impl::assign_one() {
+	const auto node_ptr = pop_assignable();
+	if(node_ptr == nullptr) return std::nullopt;
+
+	auto& node = *node_ptr;
+
+	assigned_state assigned;
+	if(const auto& eagerly_assignable = std::get_if<conditional_eagerly_assignable_state>(&node.assignment)) {
+		// After an instruction is popped from the assignment queue, "eager-assignability" is unconditional. Force-assign to the same lane to implicitly fulfill
+		// all remaining dependencies through queue ordering.
+		assigned.device = eagerly_assignable->device;
+		assigned.lane = eagerly_assignable->lane;
+		assert(
+		    assigned.lane.has_value()
+		    && get_lane_state(node.target, assigned.device, *assigned.lane).last_incomplete_submission == eagerly_assignable->expected_last_submission_on_lane);
```

Comment:
Good catch - it can be which causes the eager submission to be retracted, so it was inconsequential, but the invariant should still be that this value is never null in the eagerly-assignable state. I've added an appropriate early-return in `try_mark_for_assignment` to catch this condition early and added a test.